File: cs/ssh.py
"""`cs ssh setup|check`: per-machine SSH keys for every identity.

Naming: ~/.ssh/cs/<identity> (+ .pub), comment and GitHub title `cs:<machine>:<identity>`,
published as machines/<machine>/ssh/<identity>.pub in the config repo.

- generates missing ed25519 keys (no passphrase; disk encryption is assumed)
- publishes the public halves under machines/<machine>/ssh/ in the config repo
- writes a managed block into ~/.ssh/config
- registers pubkeys on GitHub via the owner's token when the owner is a user account,
  otherwise prints what to paste where
"""
from __future__ import annotations

import re
import subprocess
from pathlib import Path
from typing import List, Optional

from . import github, gitutil, paths, platform, ui
from .config import Machine
from .manifest import Identity, Manifest

MARK, END = "# >>> claude-share >>>", "# <<< claude-share <<<"


def _keygen(key: Path, comment: str) -> None:
    key.parent.mkdir(parents=True, exist_ok=True, mode=0o700)
    subprocess.run(["ssh-keygen", "-q", "-t", "ed25519", "-N", "", "-C", comment, "-f", str(key)], check=True)
    key.chmod(0o600)


def _github_user_for_key(key: Path) -> Optional[str]:
    p = subprocess.run(["ssh", "-T", "-i", str(key), "-o", "IdentitiesOnly=yes", "-o", "StrictHostKeyChecking=accept-new",
                        "-o", "BatchMode=yes", "git@github.com"], capture_output=True, text=True, timeout=20)
    m = re.search(r"Hi ([^!]+)!", p.stdout + p.stderr)
    return m.group(1) if m else None


def _config_block() -> str:
    lines = [MARK, "Host github.com", "    IdentitiesOnly yes", "    AddKeysToAgent yes"]
    if platform.is_macos():
        lines.append("    UseKeychain yes")
    lines.append(END)
    return "\n".join(lines) + "\n"


def write_ssh_config() -> bool:
    cfg = paths.home() / ".ssh" / "config"
    text = cfg.read_text() if cfg.exists() else ""
    block = _config_block()
    if MARK in text:
        new = text[:text.index(MARK)] + block + text[text.index(END) + len(END) + 1:]
    else:
        new = block + ("\n" if text and not text.startswith("\n") else "") + text
    if new == text:
        return False
    cfg.parent.mkdir(parents=True, exist_ok=True, mode=0o700)
    cfg.write_text(new)
    cfg.chmod(0o600)
    return True


def _register(ident: Identity, pub: str, title: str) -> str:
    owner = ident.github_owner
    if not owner:
        return "no github_owner; add the key manually"
    tok = github.get_token(owner)
    if not tok:
        return f"no token for {owner}; add manually: https://github.com/settings/ssh/new"
    try:
        if github.owner_type(owner, tok) != "User":
            return f"{owner} is an organization — add the key to the *user* account that belongs to it: https://github.com/settings/ssh/new"
        existing = github.api("GET", "/user/keys", tok)
        if any(k.get("key", "").split()[:2] == pub.split()[:2] for k in existing):
            return "already registered on GitHub"
        github.api("POST", "/user/keys", tok, {"title": title, "key": pub})
        return "registered on GitHub"
    except github.GitHubError as e:
        if " 403 " in str(e) or " 404 " in str(e):
            return f"token lacks 'Git SSH keys: write' (account permission) — add manually: https://github.com/settings/ssh/new"
        return str(e)


def setup(repo: Path, m: Machine, man: Manifest, check_only: bool = False) -> int:
    if not man.identities:
        ui.warn("no identities yet (cs identity add …)")
        return 0
    rows = []
    published = False
    for ident in man.identities.values():
        key = paths.expand(ident.key_path)
        pub_f = key.with_name(key.name + ".pub")
        state: List[str] = []
        if not key.exists():
            if check_only:
                rows.append([ident.id, ident.key_path, ui.red("missing")]); continue
            _keygen(key, f"cs:{m.name}:{ident.id}")
            state.append(ui.green("generated"))
        pub = pub_f.read_text().strip()
        dest = repo / "machines" / m.name / "ssh" / f"{ident.id}.pub"
        if not check_only and (not dest.exists() or dest.read_text().strip() != pub):
            dest.parent.mkdir(parents=True, exist_ok=True)
            dest.write_text(pub + "\n")
            gitutil.run(["add", str(dest.relative_to(repo))], repo)
            published = True
        user = _github_user_for_key(key)
        if user:
            state.append(ui.green(f"github: {user}"))
        elif not check_only:
            state.append(ui.dim(_register(ident, pub, f"cs:{m.name}:{ident.id}")))
            user = _github_user_for_key(key)
            if user:
                state[-1] = ui.green(f"github: {user}")
        else:
            state.append(ui.red("not accepted by GitHub"))
        rows.append([ident.id, ident.key_path, "  ".join(state)])
    if published:
        gitutil.commit(repo, f"machines: {m.name} ssh public keys", "cs", f"cs@{m.name}")
    if not check_only and write_ssh_config():
        ui.act("~/.ssh/config: managed block (IdentitiesOnly, AddKeysToAgent)")
    ui.table(rows, header=["identity", "key", "state"])
    unregistered = [r for r in rows if "github:" not in r[2]]
    if unregistered:
        for r in unregistered:
            key = paths.expand(r[1]).with_name(paths.expand(r[1]).name + ".pub")
            ident = man.identities.get(r[0])
            if key.exists():
                who = f"the GitHub account that belongs to {ident.github_owner}" if ident and ident.github_owner else "your GitHub account"
                ui.note(f"Add the '{r[0]}' key to {who}",
                        [ui.cyan("https://github.com/settings/ssh/new"), "", ui.bold(key.read_text().strip()), "",
                         ui.dim(f"title suggestion: cs:{m.name}:{r[0]}")])
        return 1
    return 0

File: src/ssh.ts
/** cs ssh setup|check: per-machine identity keys ~/.ssh/cs/<id>. */
import { chmodSync, existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { spawnSync } from "node:child_process";
import * as git from "./git.js";
import * as github from "./github.js";
import { expand, home, contract } from "./paths.js";
import * as platform from "./platform.js";
import type { Machine } from "./config.js";
import { keyPath, type Identity, type Manifest } from "./manifest.js";
import * as ui from "./ui.js";

const MARK = "# >>> claude-share >>>", END = "# <<< claude-share <<<";
function keygen(key: string, comment: string) { mkdirSync(dirname(key), { recursive: true, mode: 0o700 }); const p = spawnSync("ssh-keygen", ["-q", "-t", "ed25519", "-N", "", "-C", comment, "-f", key]); if (p.status !== 0) throw new Error("cs: ssh-keygen failed"); chmodSync(key, 0o600); }
export function githubUserForKey(key: string): string | undefined {
  if (process.env.CS_OFFLINE) return undefined;
  const p = spawnSync("ssh", ["-T", "-i", key, "-o", "IdentitiesOnly=yes", "-o", "StrictHostKeyChecking=accept-new", "-o", "BatchMode=yes", "git@github.com"], { encoding: "utf8", timeout: 20000, stdio: ["ignore", "pipe", "pipe"] });
  return ((p.stdout ?? "") + (p.stderr ?? "")).match(/Hi ([^!]+)!/)?.[1];
}
export function writeSshConfig(): boolean {
  const cfg = join(home(), ".ssh", "config"); const text = existsSync(cfg) ? readFileSync(cfg, "utf8") : "";
  const block = [MARK, "Host github.com", "    IdentitiesOnly yes", "    AddKeysToAgent yes", ...(platform.isMac() ? ["    UseKeychain yes"] : []), END].join("\n") + "\n";
  const next = text.includes(MARK) ? text.slice(0, text.indexOf(MARK)) + block + text.slice(text.indexOf(END) + END.length + 1) : block + (text && !text.startsWith("\n") ? "\n" : "") + text;
  if (next === text) return false; mkdirSync(dirname(cfg), { recursive: true, mode: 0o700 }); writeFileSync(cfg, next); chmodSync(cfg, 0o600); return true;
}
async function register(i: Identity, pub: string, title: string): Promise<string> {
  if (process.env.CS_OFFLINE) return "offline"; if (!i.owner) return "no owner; add the key manually";
  const tok = github.getToken(i.owner); if (!tok) return `no token for ${i.owner}; add manually: https://github.com/settings/ssh/new`;
  try { if ((await github.ownerType(i.owner, tok)) !== "User") return `${i.owner} is an organization — add the key to the user account that belongs to it: https://github.com/settings/ssh/new`;
    const keys = await github.api<any[]>("GET", "/user/keys", tok); if (keys.some((k) => (k.key ?? "").split(" ").slice(0, 2).join(" ") === pub.split(" ").slice(0, 2).join(" "))) return "already registered on GitHub";
    await github.api("POST", "/user/keys", tok, { title, key: pub }); return "registered on GitHub";
  } catch (e: any) { return /403|404/.test(e.message) ? "token lacks 'Git SSH keys: write' — add manually: https://github.com/settings/ssh/new" : e.message; }
}
export async function setup(repo: string, m: Machine, man: Manifest, checkOnly = false): Promise<number> {
  const ids = Object.values(man.identities); if (!ids.length) { ui.warn("no identities yet (cs identity add …)"); return 0; }
  const rows: string[][] = []; let published = false; const unregistered: Identity[] = [];
  for (const i of ids) {
    const key = expand(keyPath(i)), pubf = key + ".pub"; const state: string[] = [];
    if (!existsSync(key)) { if (checkOnly) { rows.push([i.id, keyPath(i), ui.red("missing")]); unregistered.push(i); continue; } keygen(key, `cs:${m.name}:${i.id}`); state.push(ui.green("generated")); }
    const pub = readFileSync(pubf, "utf8").trim(); const dest = join(repo, "machines", m.name, "ssh", `${i.id}.pub`);
    if (!checkOnly && (!existsSync(dest) || readFileSync(dest, "utf8").trim() !== pub)) { mkdirSync(dirname(dest), { recursive: true }); writeFileSync(dest, pub + "\n"); git.git(["add", dest], repo); published = true; }
    let user = await ui.spin(`verifying ${i.id} key on GitHub…`, async () => githubUserForKey(key));
    if (user) state.push(ui.green(`github: ${user}`));
    else if (!checkOnly) { const r = await ui.spin(`registering ${i.id} key…`, async () => register(i, pub, `cs:${m.name}:${i.id}`)); user = githubUserForKey(key); if (user) state.push(ui.green(`github: ${user}`)); else { state.push(ui.yellow(r.startsWith("registered") ? "registered, not verified yet" : "needs registering")); unregistered.push(i); } }
    else { state.push(ui.yellow("not accepted by GitHub yet")); unregistered.push(i); }
    rows.push([i.id, keyPath(i), state.join("  ")]);
  }
  if (published) git.commit(repo, `machines: ${m.name} ssh public keys`, "cs", `cs@${m.name}`);
  if (!checkOnly && writeSshConfig()) ui.step("~/.ssh/config: managed block (IdentitiesOnly, AddKeysToAgent)");
  ui.table(rows, ["identity", "key", "state"]);
  for (const i of unregistered) { const pubf = expand(keyPath(i)) + ".pub"; if (!existsSync(pubf)) continue;
    const who = i.owner && i.owner.toLowerCase() !== i.id.toLowerCase() ? `the ${i.owner} account` : `your GitHub account that is a member of ${i.owner || "the org"}`;
    ui.note([`${ui.cyan("https://github.com/settings/ssh/new")}  ${ui.dim(`→ logged in as ${who}`)}`, "", `title  ${ui.bold(`cs:${m.name}:${i.id}`)}`, `key    ${ui.bold(readFileSync(pubf, "utf8").trim())}`], `Add the ${i.id} key`); }
  return unregistered.length ? 1 : 0;
}
